File: moodstocks-api/msapi.m
/**
 * Usage notes
 * --
 * This source code is not intended to be used in a standalone fashion. You
 * should cherry pick and adapt it into your iOS/Cocoa project, e.g.:
 *
 *   In practice you should *always* use asynchronous requests and define a
 *   delegate (in general the current controller of your application)
 *   The sample below performs synchronous requests on purpose, i.e. to ensure
 *   sequential operations add > search > remove
 *   See http://allseeing-i.com/ASIHTTPRequest/How-to-use for more details
 *
 * Pre-requisites
 * --
 *   ASIHTTPRequest
 *   You should use a recent version of ASIHTTPRequest since issues related to
 *   Digest Authentication from v1.8 have been fixed (e.g. these examples have
 *   been succesfully tested with commit 4b3eb59620eaa32eed98 from
 *   https://github.com/Moodstocks/asi-http-request
 *
 *   json-framework
 *   See https://github.com/stig/json-framework
 */
 
#import "JSON.h"
#import "ASIHTTPRequest.h"
#import "ASIFormDataRequest.h"

static NSString* kMSAPIKey            = @"kEy";
static NSString* kMSAPISecret         = @"sEcReT";
static NSString* kMSImageFilename     = @"sample.jpg";
static NSString* kMSImageURL          = @"http://api.moodstocks.com/static/sample-book.jpg";
static NSString* kMSID                = @"test1234";
static NSString* kMSAPIEchoURL        = @"http://api.moodstocks.com/v2/echo";
static NSString* kMSAPIRefURL         = @"http://api.moodstocks.com/v2/ref";
static NSString* kMSAPISearchURL      = @"http://api.moodstocks.com/v2/search";

void MSDisp(ASIHTTPRequest* request) {
    NSString* context = [[request userInfo] objectForKey:@"context"];
    NSString* json = [[NSString alloc] initWithData:[request responseData]
                                           encoding:NSUTF8StringEncoding];
    NSDictionary* resp = (NSDictionary*) [json JSONValue];
    NSLog(@"%@: %@", context, [resp description]);
    [json release];
}

void MSErrDisp(ASIHTTPRequest* request) {
    NSString* context = [[request userInfo] objectForKey:@"context"];

    NSLog(@"[%@] REQUEST FAILED, code: %d", context, [[request error] code]);
}

int main(int argc, char *argv[]) {
    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];

    /**
     * Authenticating with your API key (Echo service)
     */
    {
        NSString* echoURL = [kMSAPIEchoURL stringByAppendingString:@"?foo=bar&bacon=chunky"];
        ASIHTTPRequest* request = [ASIHTTPRequest requestWithURL:[NSURL URLWithString:echoURL]];
        [request setUserInfo:[NSDictionary dictionaryWithObject:@"echo" forKey:@"context"]];
        [request setAuthenticationScheme:(NSString *) kCFHTTPAuthenticationSchemeDigest];
        [request setUsername:kMSAPIKey];
        [request setPassword:kMSAPISecret];
        [request startSynchronous];
        NSError* error = [request error];
        if (!error)
            MSDisp(request);
        else
            MSErrDisp(request);
    } // end of echo


    /**
     * Adding objects to recognize
     */
    UIImage* image = [UIImage imageNamed:kMSImageFilename];
    NSData *imageData = UIImageJPEGRepresentation(image, 0.75);
    NSString* objectURL = [kMSAPIRefURL stringByAppendingFormat:@"/%@", kMSID];
    {
        ASIFormDataRequest* request = [ASIFormDataRequest requestWithURL:[NSURL URLWithString:objectURL]];
        [request setUserInfo:[NSDictionary dictionaryWithObject:@"add" forKey:@"context"]];
        [request setRequestMethod:@"PUT"];
        [request setData:imageData forKey:@"image_file"];
        [request setAuthenticationScheme:(NSString *) kCFHTTPAuthenticationSchemeDigest];
        [request setUsername:kMSAPIKey];
        [request setPassword:kMSAPISecret];
        [request startSynchronous];
        NSError* error = [request error];
        if (!error)
            MSDisp(request);
        else
            MSErrDisp(request);
    } // end of add

    /**
     * Looking up objects
     */
    {
        ASIFormDataRequest* request = [ASIFormDataRequest requestWithURL:[NSURL URLWithString:kMSAPISearchURL]];
        [request setUserInfo:[NSDictionary dictionaryWithObject:@"search" forKey:@"context"]];
        [request setRequestMethod:@"POST"];
        [request setData:imageData forKey:@"image_file"];
        [request setAuthenticationScheme:(NSString *) kCFHTTPAuthenticationSchemeDigest];
        [request setUsername:kMSAPIKey];
        [request setPassword:kMSAPISecret];
        [request startSynchronous];
        NSError* error = [request error];
        if (!error)
            MSDisp(request);
        else
            MSErrDisp(request);
    } // end of search

    /**
     * Updating a reference & using a hosted image
     */
    {
        ASIFormDataRequest* request = [ASIFormDataRequest requestWithURL:[NSURL URLWithString:objectURL]];
        [request setUserInfo:[NSDictionary dictionaryWithObject:@"update" forKey:@"context"]]; 
        [request setRequestMethod:@"PUT"];
        [request setPostValue:kMSImageURL forKey:@"image_url"];
        [request setAuthenticationScheme:(NSString *) kCFHTTPAuthenticationSchemeDigest];
        [request setUsername:kMSAPIKey];
        [request setPassword:kMSAPISecret];    
        [request startSynchronous];
        NSError* error = [request error];
        if (!error)
            MSDisp(request);
        else
            MSErrDisp(request);
    } // end of update with a hosted image

    /**
     * Removing reference images
     */
    {
        ASIHTTPRequest* request = [ASIHTTPRequest requestWithURL:[NSURL URLWithString:objectURL]];
        [request setUserInfo:[NSDictionary dictionaryWithObject:@"remove" forKey:@"context"]];
        [request setRequestMethod:@"DELETE"];
        [request setAuthenticationScheme:(NSString *) kCFHTTPAuthenticationSchemeDigest];
        [request setUsername:kMSAPIKey];
        [request setPassword:kMSAPISecret];
        [request startSynchronous];
        NSError* error = [request error];
        if (!error)
            MSDisp(request);
        else
            MSErrDisp(request);
    } // end of remove

    [pool release];
    return 0;
}

File: moodstocks-api/msapi.c
/* Compile with -lcurl */

#include <stdio.h>
#include <string.h>
#include <curl/curl.h>
#include <curl/types.h>
#include <curl/easy.h>

size_t static disp(void *buffer,size_t size,size_t nmemb,void *userp){
  printf("%s\n",(char *)buffer);
  return (size_t)0;
}

int main(int argc,char **argv){
  
  /* Settings */
  char *key = "YourApiKey";
  char *secret = "YourApiSecret";
  char *image_filename = "sample.jpg";
  char *image_url = "http://api.moodstocks.com/static/sample-book.jpg";
  char *id = "test1234";

  char *api_ep = "http://api.moodstocks.com/v2";
  char kpw[40]; sprintf(kpw,"%s:%s",key,secret);
  char url[256];

  CURL *curl;
  CURLcode res;
  curl_global_init(CURL_GLOBAL_ALL);

  struct curl_httppost *formpost = NULL;
  struct curl_httppost *lastptr = NULL;
  curl_formadd( &formpost,&lastptr,
                CURLFORM_COPYNAME,"image_file",
                CURLFORM_FILE,image_filename,
                CURLFORM_END );

  struct curl_httppost *formpost2 = NULL;
  struct curl_httppost *lastptr2 = NULL;
  curl_formadd( &formpost2,&lastptr2,
                CURLFORM_COPYNAME,"image_url",
                CURLFORM_COPYCONTENTS,image_url,
                CURLFORM_END );

  /* Authenticating with your API key (Echo service) */

  strcpy(url,api_ep); strcat(url,"/echo?foo=bar");

  curl = curl_easy_init();
  curl_easy_setopt(curl,CURLOPT_HTTPAUTH,CURLAUTH_DIGEST);
  curl_easy_setopt(curl,CURLOPT_URL,url);
  curl_easy_setopt(curl,CURLOPT_USERPWD,kpw);
  curl_easy_setopt(curl,CURLOPT_WRITEFUNCTION,disp);

  curl_easy_perform(curl);
  curl_easy_cleanup(curl);

  /* Adding objects to recognize */
  
  strcpy(url,api_ep); strcat(url,"/ref/"); strcat(url,id);

  curl = curl_easy_init();
  curl_easy_setopt(curl,CURLOPT_HTTPAUTH,CURLAUTH_DIGEST);
  curl_easy_setopt(curl,CURLOPT_URL,url);
  curl_easy_setopt(curl,CURLOPT_USERPWD,kpw);
  curl_easy_setopt(curl,CURLOPT_WRITEFUNCTION,disp);

  curl_easy_setopt(curl,CURLOPT_HTTPPOST,formpost);
  curl_easy_setopt(curl,CURLOPT_CUSTOMREQUEST,"PUT");

  curl_easy_perform(curl);
  curl_easy_cleanup(curl);

  /* Looking up objects */

  strcpy(url,api_ep); strcat(url,"/search");

  curl = curl_easy_init();
  curl_easy_setopt(curl,CURLOPT_HTTPAUTH,CURLAUTH_DIGEST);
  curl_easy_setopt(curl,CURLOPT_URL,url);
  curl_easy_setopt(curl,CURLOPT_USERPWD,kpw);
  curl_easy_setopt(curl,CURLOPT_WRITEFUNCTION,disp);

  curl_easy_setopt(curl,CURLOPT_HTTPPOST,formpost);

  curl_easy_perform(curl);
  curl_easy_cleanup(curl);

  /* Updating a reference & using a hosted image */

  strcpy(url,api_ep); strcat(url,"/ref/"); strcat(url,id);

  curl = curl_easy_init();
  curl_easy_setopt(curl,CURLOPT_HTTPAUTH,CURLAUTH_DIGEST);
  curl_easy_setopt(curl,CURLOPT_URL,url);
  curl_easy_setopt(curl,CURLOPT_USERPWD,kpw);
  curl_easy_setopt(curl,CURLOPT_WRITEFUNCTION,disp);

  curl_easy_setopt(curl,CURLOPT_HTTPPOST,formpost2);
  curl_easy_setopt(curl,CURLOPT_CUSTOMREQUEST,"PUT");

  curl_easy_perform(curl);
  curl_easy_cleanup(curl);

  /* Removing reference images */

  strcpy(url,api_ep); strcat(url,"/ref/"); strcat(url,id);

  curl = curl_easy_init();
  curl_easy_setopt(curl,CURLOPT_HTTPAUTH,CURLAUTH_DIGEST);
  curl_easy_setopt(curl,CURLOPT_URL,url);
  curl_easy_setopt(curl,CURLOPT_USERPWD,kpw);
  curl_easy_setopt(curl,CURLOPT_WRITEFUNCTION,disp);

  curl_easy_setopt(curl,CURLOPT_CUSTOMREQUEST,"DELETE");

  curl_easy_perform(curl);
  curl_easy_cleanup(curl);

  /* cleanup */
  curl_formfree(formpost);
  return 0;

}
